### PR TITLE
adding has_many_remote

### PR DESCRIPTION
**Pull request by rilian**

### CODE REVIEW

**Comment by denyago** [comment]
Thanks for pull!

`.all` was added with the intention, not to fetch unnecessary remote objects. Say, you have this query:

``` ruby
# There are 2 users with names 'User A' and 'User B'...
User.where(t[:name].matches('%User%')).includes_remote(:profile).
        where(t[:name].matches('%B%'))
```

`User.where(t[:name].matches('%User%'))` will have association for 2 users, remote_association will load objects for both. Then, `where(t[:name].matches('%B%'))` will remove one user from collection. So remote_association done unnecessary job. And if we were dealing with 100K records on step 1 and 10 records on step 2?

The idea, is to do a lazy-load of remote objects. We need to know, when ActiveRecord loaded data from relation to array.

This test should pass, then:

``` ruby
  it "should fetch remote objects right after ActiveRecord fetched array of data" do
    FakeWeb.register_uri(:get, "#{REMOTE_HOST}/profiles.json?user_id%5B%5D=2", 
       body: [{profile: {id: 2, user_id: 2, like: "letter B"}}].to_json)

    t = User.arel_table
    users = User.where(t[:name].matches('%User%')).includes_remote(:profile).
                         where(t[:name].matches('%B%')).all
    users.map(&:profile).flatten.map(&:like).should eq(['letter B'])
  end


```

Patching of this, may help.... https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/relation.rb#L150

**Comment by rilian** [comment]
we need sorting and filtering based on values of remote objects.

**Comment by rilian** [comment]
as from discussion:

1) We do use the `includes_remote()` at the end of ActiveRecord query and keep it returning Array
2) For controller cases, when we need sorting (ordering) based on remote data, and then pagination, we implement custom methods for `order`, `limit`, `offset`

File: lib/remote_association/active_record/relation.rb [hunk 1]
```diff
@@ -33,37 +34,47 @@ def includes_remote(*args)
 
     private
 
-      def fetch_and_join_for_has_one_remote(ar_accessor, foreign_key, ar_class)
-        keys = self.uniq.pluck(:id)
+    def fetch_and_join_for_has_one_remote(ar_accessor, foreign_key, ar_class)
```
Comment (denyago) [inline lib/remote_association/active_record/relation.rb:37]: RubyMine autoformating sucks: I have `private` right under `class`.

I hate programs, smarter, than me.

**Comment by Lightpower** [comment]
Looks good

File: lib/remote_association/has_many_remote.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  module HasManyRemote
+    # Specifies a one-to-many association with another class. This method should only be used
+    # if this class is a kind of ActiveResource::Base and service for this resource can
+    # return some kind of foreign key.
+    #
+    # Methods will be added for retrieval and query for a single associated object, for which
+    # this object holds an id:
+    #
+    # [associations()]
+    #   Returns the associated objects. +nil+ is returned if none is found.
```
Comment (denyago) [inline lib/remote_association/has_many_remote.rb:11]: In this case, it should return empty array.

Comment (rilian) [inline lib/remote_association/has_many_remote.rb:11]: yes, really

File: lib/remote_association/has_many_remote.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+      add_activeresource_relation(remote_rel.to_sym, rel_options)
+
+      class_eval <<-RUBY, __FILE__, __LINE__+1
+
+        attr_accessor :#{remote_rel}
+
+        def #{remote_rel}                                                 #  def customers
+          if remote_resources_prefetched?                                 #    if remote_resources_prefetched?
+            @#{remote_rel} ? @#{remote_rel} : nil                         #      @customers ? @customers : nil
```
Comment (denyago) [inline lib/remote_association/has_many_remote.rb:55]: Array instead of nil

**Pull request merged**